include/bluetooth/bluetooth.h

```diff
@@ -398,6 +463,118 @@ int bt_le_adv_update_data(const struct bt_data *ad, size_t ad_len,
  */
 int bt_le_adv_stop(void);
 
+/** @brief Increment an advertising set`s reference count.
+ *
+ *  Increment the reference count of an advertising set object.
+ *
+ *  @param adv Advertising set object.
+ *
+ *  @return Advertising set object with incremented reference count.
+ */
+struct bt_adv *bt_adv_ref(struct bt_adv *adv);
+
+/** @brief Decrement an advertising set's reference count.
+ *
+ *  Decrement the reference count of an advertising setobject.
+ *
+ *  @param adv Advertising set object.
+ */
+void bt_adv_unref(struct bt_adv *adv);
+
+/** @brief Create advertising set.
+ *
+ *  Create a new advertising set and set advertising parameters.
+ *
+ *  @param param Advertising parameters.
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ *  @return -ECONNREFUSED When connectable advertising is requested and there
+ *			  is already maximum number of connections established.
+ *			  This error code is only guaranteed when using Zephyr
+ *			  controller, for other controllers code returned in
+ *			  this case may be -EIO.
+ */
+struct bt_adv *bt_le_adv_set_create(const struct bt_le_adv_param *param);
```

Comment:
So the idea here is that we would do memory allocation? I rather not do this since we may have different parts trying to advertise so application would only notice the lack of memory at runtime, so I rather leave the memory allocation to the application directly.

Reply:
@Vudentz It would allocate one of a fixed number of advertising contexts in the host. This one is quite small and mostly just has the advertising_handle and some state variables. This API would create a new advertising context in the controller. And it is not possible to give allocated memory to the controller over the HCI interface.

This is also similar to how conn objects are allocated. This would be symmetrical to bt_conn_create_le (which wraps the connection handle plus internal state).

Reply:
So we are assuming we know the number of sets available in the controller? How about split builds? Either way we cannot mix legacy advertisement with extended ones over HCI as that would fail so having an API which masks this sort of detail is probably desirable, we could make the sets work even without the controller supporting it by doing the advertising set context on the host, anyway the spec don't seem to describe how the scheduling should happen so just rotating the sets should work.

Reply:
> So we are assuming we know the number of sets available in the controller?
> How about split builds?

That's a good question actually. The number of advertising sets in the controller can be read using
LE Read Number of Supported Advertising Sets command. However it has this note:

	Note: The number of advertising sets that can be supported is not fixed and the
	Controller can change it at any time because the memory used to store
	advertising sets can also be used for other purposes.

I think we can assume in combined build that we would have a specific number of advertising sets set by Kconfig.
In the split build case, I think we should just make sure that we atleast have enough advertising set contexts in the host.
The context itself is not very large so i don't see this is a problem.
In any case the larger memory usage is in the HCI both for advertising context and advertising data.
In split build we cannot assign memory to the controller, maybe in the combined case, but that would still have to be something that is outside of HCI.

> Either way we cannot mix legacy advertisement with extended ones over HCI as that would fail so having an API which masks this sort of detail is probably desirable,

We cannot mix legacy and extended, but we can support the old advertising API with the new API, limited to one set.
And the old advertising API can use the new advertising HCI commands, with the added advantage of solving the multiple identities problem, solving a structural issue with the current API.

>  we could make the sets work even without the controller supporting it by doing the advertising set context on the host, anyway the spec don't seem to describe how the scheduling should happen so just rotating the sets should work.

In this case we would have to rotate advertising parameters and advertising data, so then we would either have to copy them in the host, or require the application
to provide us pointers to memory that stays valid while for the lifetime of the advertising set.
In this case I think it would maybe be easier to keep with the old recommendation of cycling the advertising data of a single advertiser instead.